### PR TITLE
fix: support arbitrary values for `checkpointing.metric_name`

### LINKED ISSUE #1261: config option checkpointing.metric_name is not respected
The `checkpointing.metric_name` option is not respected and instead the SFT/GRPO/other trainers always default to "val_loss". Relevant code here: https://github.com/NVIDIA-NeMo/RL/blob/376e6254d11cc6ef5378f784dd5e05cc8309d7b8/nemo_rl/algorithms/sft.py#L497-L510.

Likely need to check the `val_metrics` dict for `metric_name` and issue a warning there instead of checking `sft_save_state`.

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?

**Add a one line overview of what this PR aims to accomplish.**

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

closes #1261 

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

File: examples/configs/distillation_math.yaml [hunk 1]
```diff
@@ -19,7 +19,7 @@ loss_fn:
 checkpointing:
     enabled: true
     checkpoint_dir: "checkpoints/distillation-${policy.model_name}"
-    metric_name: "val_reward"
+    metric_name: "val:reward"
```
Comment (samodi-nv) [inline examples/configs/distillation_math.yaml]: We should document that the options are "train:" or "val:" in the config python code (`checkpoint.py::CheckpointingConfig`)

File: nemo_rl/algorithms/sft.py [hunk 2]
```diff
@@ -508,24 +508,40 @@ def sft_train(
                 if master_config["checkpointing"]["enabled"] and (
                     should_save_by_step or should_save_by_timeout
                 ):
+
+                    original_metric_name = master_config["checkpointing"]["metric_name"]
```
Comment (samodi-nv) [inline nemo_rl/algorithms/sft.py]:
```suggestion
                    full_save_metric_name = master_config["checkpointing"]["metric_name"]
```

nit for clarity

File: nemo_rl/algorithms/sft.py [hunk 3]
```diff
@@ -508,24 +508,40 @@ def sft_train(
...
-                    if master_config["checkpointing"]["metric_name"] is not None:
-                        if (
-                            master_config["checkpointing"]["metric_name"]
-                            not in sft_save_state
-                        ):
-                            warnings.warn(
-                                f"You asked to save checkpoints based on {master_config['checkpointing']['metric_name']} but the metric is not found in the save state. "
-                                "This checkpoint will not be saved as top-k."
-                            )
+                    if original_metric_name is not None:
```
Comment (samodi-nv) [inline nemo_rl/algorithms/sft.py:523]:
```suggestion
                if original_metric_name is not None:
                    metrics_source = metrics if train_or_val == "train" else val_metrics
                    if metric_name not in metrics_source:
                        warnings.warn(
                            f"You asked to save checkpoints based on {metric_name} but the metric is not found in the {train_or_val} metrics. "
                            "This checkpoint will not be saved as top-k."
                        )
                        if original_metric_name in sft_save_state:
                            del sft_save_state[original_metric_name]
                    else:
                        sft_save_state[original_metric_name] = metrics_source[metric_name]
    
```

nit: can make this more concise. same comment for the other algorithms.

**Review by samodi-nv — CHANGES_REQUESTED** [review]

File: nemo_rl/algorithms/sft.py [hunk 4]
```diff
@@ -508,24 +508,40 @@ def sft_train(
...
+                        else:
+                            if val_metrics is None or metric_name not in val_metrics:
+                                warnings.warn(
+                                    f"You asked to save checkpoints based on {metric_name} but the metric is not found in the validation metrics. "
+                                    "This checkpoint will not be saved as top-k."
+                                )
+                                if original_metric_name in sft_save_state:
+                                    del sft_save_state[original_metric_name]
+                            else:
+                                sft_save_state[original_metric_name] = val_metrics[metric_name]
```
Comment (samodi-nv) [inline nemo_rl/algorithms/sft.py]:
```suggestion
                if original_metric_name is not None:
                    metrics_source = metrics if train_or_val == "train" else val_metrics
                    if metric_name not in metrics_source:
                        warnings.warn(
                            f"You asked to save checkpoints based on {metric_name} but the metric is not found in the {train_or_val} metrics. "
                            "This checkpoint will not be saved as top-k."
                        )
                        if original_metric_name in sft_save_state:
                            del sft_save_state[original_metric_name]
                    else:
                        sft_save_state[original_metric_name] = metrics_source[metric_name]
    
```
nit: can make this more concise. same comment for the other algorithms.

File: nemo_rl/algorithms/distillation.py [hunk 5]
```diff
@@ -731,17 +731,30 @@ def distillation_train(
...
-                    if master_config["checkpointing"]["metric_name"] is not None:
-                        if (
-                            master_config["checkpointing"]["metric_name"]
-                            not in distillation_save_state
-                        ):
+                    full_metric_name = master_config["checkpointing"]["metric_name"]
+                    if full_metric_name is not None:
+                        assert full_metric_name.count(":") == 1, (
+                            "metric_name must contain exactly one colon"
+                        )
```
Comment (terrykong) [inline nemo_rl/algorithms/distillation.py]: maybe a more descriptive error since there are more constraints than just `:`. disclaimer, i didn't test or lint this, but maybe something like this
```suggestion
                        assert full_metric_name.startswith('train:') or full_metric_name.startswith('val:'), (
                            f"metric_name={full_metric_name} must start with 'val:' or 'train:'"
                        )
```

same with the other algos

File: nemo_rl/algorithms/distillation.py [hunk 6]
```diff
@@ -731,17 +731,32 @@ def distillation_train(
...
+                        parts = full_metric_name.split(":")
+                        train_or_val = "val" if "val" in parts[0] else "train"
+                        metric_name = parts[1]
+
+                        metrics_source = (
+                            metrics if train_or_val == "train" else val_metrics
+                        )
+                        if not metrics_source or metric_name not in metrics_source:
                             warnings.warn(
```
Comment (terrykong) [inline nemo_rl/algorithms/distillation.py:807]: seeing this `"val:validation-default_loss"` makes me think we should error loudly as opposed to just warn to avoid people thinking their checkpoint is being saved according to val loss. If it's set but their metrics don't include it, then the error message can just say to set it to `null`, which the intent would be conveyed when someone looks at the saved config.yaml

Comment (ashors1) [reply nemo_rl/algorithms/distillation.py:807]: Understood. The reason we're not erroring is because we want to enable the case where checkpoint_period % val_period != 0. In this case, only some of the checkpoint steps would have associated validation metrics. However, I think we could error in the case that validation metrics _are_ present and the key still isn't present in the metrics dict.

**Review by terrykong — APPROVED** [review]
